src/main/java/com/imgix/URLBuilder.java

```diff
@@ -73,72 +84,241 @@ public String createURL(String path, Map<String, String> params) {
+    }
 
-        for (int ratio: srcsetTargetRatios) {
-            params.put("dpr", Integer.toString(ratio));
-            srcset += this.createURL(path, params) + " " + ratio + "x,\n";
-        }
+    private String createSrcSetDPR(String path, Map<String, String> params, boolean disableVariableQuality) {
+        StringBuilder srcset = new StringBuilder();
+        Map<String, String> srcsetParams = new HashMap<String, String>(params);
 
+        boolean has_quality = params.get("q") != null;
+
+        for (int ratio: TARGET_RATIOS) {
+            srcsetParams.put("dpr", Integer.toString(ratio));
+
+            if (!disableVariableQuality && !has_quality) {
+                srcsetParams.put("q", DPR_QUALITIES[ratio - 1].toString());
+            }
+            srcset.append(this.createURL(path, srcsetParams)).append(" ").append(ratio).append("x,\n");
+        }
         return srcset.substring(0, srcset.length() - 2);
     }
 
-    private static ArrayList<Integer> targetWidths() {
+    /**
+     * Create an `ArrayList` of integer target widths.
+     *
+     * If `begin`, `end`, and `tol` are the default values, then the
+     * targets are not custom, in which case the default widths are
+     * returned.
+     *
+     * A target width list of length one is valid: if `begin` == `end`
+     * then the list begins where it ends.
+     *
+     * When the `while` loop terminates, `begin` is greater than `end`
+     * (where `end` less than or equal to `MAX_WIDTH`). This means that
+     * the most recently appended value may, or may not, be the `end`
+     * value.
+     *
+     * To be inclusive of the ending value, we check for this case and the
+     * value is added if necessary.
+     *
+     * @param begin - beginning image width value
+     * @param end - ending image width value
+     * @param tol - tolerable amount of width value variation
+     * @return array list of image width values
+     */
+    public static ArrayList<Integer> targetWidths(int begin, int end, int tol) {
+        return computeTargetWidths(begin, end, tol);
+    }
+
+    public static ArrayList<Integer> targetWidths() {
+        return new ArrayList<Integer>(Arrays.asList(SRCSET_TARGET_WIDTHS));
+    }
+
+    /**
+     * Create an `ArrayList` of integer target widths.
+     *
+     * This function is the implementation details of `targetWidths`.
+     * This function exists to provide a consistent interface for
+     * callers of `targetWidths`.
+     *
+     * This function implements the syntax that fulfills the semantics
+     * of `targetWidths`. Meaning, `begin`, `end`, and `tol` are
+     * to be whole integers, but computation requires `double`s. This
+     * function hides this detail from callers.
+     */
+    private static ArrayList<Integer> computeTargetWidths(double begin, double end, double tol) {
+        if (notCustom(begin, end, tol)) {
+            return new ArrayList<Integer>(Arrays.asList(SRCSET_TARGET_WIDTHS));
```

Comment:
Could we dry this up by using `targetWidths()`?
```suggestion
            return targetWidths();
```

Reply:
Yeah man, if you want! I just didn't do that the first time because the logic felt a little circular, they're one-liners to begin with, and I wanted things to somewhat bottom out in the `private computeTargetWidths` function. But yeah, if you think it's drier then I'll push the new commit.

Reply:
That's a good point, I'll defer to you then 😁